autonat.go

```diff
@@ -75,198 +70,222 @@ func NewAutoNAT(ctx context.Context, h host.Host, getAddrs GetAddrs) AutoNAT {
+				as.lastInbound = time.Now()
+			}
 
-		delay := AutoNATRefreshInterval
-		if as.status == NATStatusUnknown {
-			delay = AutoNATRetryInterval
-		}
+		case <-netChangeChan:
+			if as.confidence > 1 {
+				as.confidence--
+			}
 
-		select {
+		// probe finished.
+		case result, ok := <-as.observations:
+			if !ok {
+				return
+			}
+			as.recordObservation(result)
 		case <-time.After(delay):
 		case <-as.ctx.Done():
 			return
 		}
-	}
-}
-
-func (as *AmbientAutoNAT) autodetect() {
-	peers := as.getPeers()
-
-	if len(peers) == 0 {
-		log.Debugf("skipping NAT auto detection; no autonat peers")
-		return
-	}
-
-	cli := NewAutoNATClient(as.host, as.getAddrs)
-	ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
-	defer cancel()
 
-	var result struct {
-		sync.Mutex
-		private int
-		public  int
-		pubaddr ma.Multiaddr
+		delay = as.scheduleProbe()
 	}
+}
 
-	probe := 3 - as.confidence
-	if probe == 0 {
-		probe = 1
-	}
-	if probe > len(peers) {
-		probe = len(peers)
+// scheduleProbe calculates when the next probe should be scheduled for,
+// and launches it if that time is now.
+func (as *AmbientAutoNAT) scheduleProbe() time.Duration {
+	// Our baseline is a probe every 'AutoNATRefreshInterval'
+	// This is modulated by:
+	// * recent inbound connections make us willing to wait up to 2x longer between probes.
+	// * low confidence makes us speed up between probes.
+	fixedNow := time.Now()
+	currentStatus := as.status.Load().(autoNATResult)
+
+	nextProbe := fixedNow
+	if !as.lastProbe.IsZero() {
+		untilNext := AutoNATRefreshInterval
+		if currentStatus.Reachability == network.ReachabilityUnknown {
+			untilNext = AutoNATRetryInterval
+		} else if currentStatus.Reachability == network.ReachabilityPublic && as.lastInbound.After(as.lastProbe) {
+			untilNext *= 2
+		} else if as.confidence < 3 {
+			untilNext = AutoNATRetryInterval
+		}
+		nextProbe = as.lastProbe.Add(untilNext)
 	}
-
-	var wg sync.WaitGroup
-
-	for _, pi := range peers[:probe] {
-		wg.Add(1)
-		go func(pi peer.AddrInfo) {
-			defer wg.Done()
-
-			as.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, peerstore.TempAddrTTL)
-			a, err := cli.DialBack(ctx, pi.ID)
-
-			switch {
-			case err == nil:
-				log.Debugf("Dialback through %s successful; public address is %s", pi.ID.Pretty(), a.String())
-				result.Lock()
-				result.public++
-				result.pubaddr = a
-				result.Unlock()
-
-			case IsDialError(err):
-				log.Debugf("Dialback through %s failed", pi.ID.Pretty())
-				result.Lock()
-				result.private++
-				result.Unlock()
-
-			default:
-				log.Debugf("Dialback error through %s: %s", pi.ID.Pretty(), err)
-			}
-		}(pi)
+	if fixedNow.After(nextProbe) || fixedNow == nextProbe {
+		go as.probeNextPeer()
+		return AutoNATRetryInterval
 	}
+	return nextProbe.Sub(fixedNow)
+}
 
-	wg.Wait()
-
-	as.mx.Lock()
-	if result.public > 0 {
+// Update the current status based on an observed result.
+func (as *AmbientAutoNAT) recordObservation(observation autoNATResult) {
```

Comment:
@willscott 

Considering the  nature of the code in this function, please can you write table unit tests for this ?